### PR TITLE
Lazy uuid parsing

### PR DESCRIPTION
**Pull request by aztech-dev**
Improves performance when generating UUID's via the PECL UUID extension by delaying parsing of the uuid string until necessary.

Refs #56

### CODE REVIEW

**Comment by jaymecd** [comment]
@aztech-dev, not sure if it's a good idea to create another class `LazyUuid`, which overwrites majority methods, as it may bring doubts for implementation use cases and especially when developer expects `LazyUuid` instance in his method by mistake.

Think it would be fine to implement this patch inside `Uuid` class directly.

**Comment by aztech-dev** [comment]
Not too sure also, but not convinced Uuid is the right place either. Can you clarify your point about doubts for implementation ? I'm not sure I understand what you mean... Thanks :)

File: src/LazyUuid.php [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        parent::__construct(array(), $numberConverter, $codec);
+    }
+
+    private function parse()
+    {
+        if ($this->isParsed) {
+            return;
+        }
+
+        $this->fields = $this->codec->decode($this->uuidString)->getFieldsHex();
```
Comment (aztech-dev) [inline src/LazyUuid.php]: Missing assignment : $this->isParsed = true

**Comment by jaymecd** [comment]
Ah, my bad :) Completely forgot about `UuidInterface`. 
Anyway, my explanation - normally you rely on a contract (interface), but `Uuid` is a class-as-a-contract (**before 3.0**). And now we bring `LazyUuid` class-as-a-contract.
~~Thats my problem~~
U could skip this comment, hard Monday :)

**Comment by aztech-dev** [comment]
Ok :) Good thing I asked... but LazyUuid extends Uuid (most people are type-hinted against Uuid, not UuidInterface, so I felt it was safer) anyways, so it should not matter.

**Comment by ramsey** [comment]
I'm also concerned about `LazyUuid`. Should we lazy-parse always for the performance gain, or are we only getting the performance benefit of lazy parsing when we're using `PeclUuid`?

**Comment by aztech-dev** [comment]
I originally went for the 'always lazy parse' approach by returning a LazyUuid the `PeclUuidFactory::fromString()`, instead of returning a LazyUuid from the `uuid1()` and `uuid4()` methods. However, some unit tests for `fromString()` were failing since you can get by with passing an invalid/malformed string, so I did not keep the behavior. I'll try to push a smoother implementation in a bit.

**Comment by ramsey** [comment]
I've done a bunch of merging around and making changes, so this is out of sync with master. You may want to rebase or submit a new PR.

**Comment by ramsey** [comment]
@aztech-dev I'm getting ready to release a 3.0.0-alpha1. Is there any chance you can take a look and rebase this against master this week? If not, I can give it a try.

**Comment by aztech-dev** [comment]
Sure, consider it done

**Comment by aztech-dev** [comment]
I've rebased and fixed the (almost) minor issue that was laying around. However, I'm still not too sure about how to handle the lazy parsing. One hand, it could be nice to always lazy parse, on the other, lazy parsing skips validation of given string. Thoughts ?

**Comment by aztech-dev** [comment]
I'm closing this in favor of #68 which feels like a better/cleaner option than this one.

**Pull request closed without merging**